// antd/lib/timeline/TimelineItem.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var React = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var __rest = void 0 && (void 0).__rest || function (s, e) {
  var t = {};

  for (var p in s) {
    if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0) t[p] = s[p];
  }

  if (s != null && typeof Object.getOwnPropertySymbols === "function") for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
    if (e.indexOf(p[i]) < 0) t[p[i]] = s[p[i]];
  }
  return t;
};

var TimelineItem =
/*#__PURE__*/
function (_React$Component) {
  (0, _inherits2["default"])(TimelineItem, _React$Component);

  function TimelineItem() {
    (0, _classCallCheck2["default"])(this, TimelineItem);
    return (0, _possibleConstructorReturn2["default"])(this, (0, _getPrototypeOf2["default"])(TimelineItem).apply(this, arguments));
  }

  (0, _createClass2["default"])(TimelineItem, [{
    key: "render",
    value: function render() {
      var _classNames, _classNames2;

      var _a = this.props,
          prefixCls = _a.prefixCls,
          className = _a.className,
          _a$color = _a.color,
          color = _a$color === void 0 ? '' : _a$color,
          children = _a.children,
          pending = _a.pending,
          dot = _a.dot,
          restProps = __rest(_a, ["prefixCls", "className", "color", "children", "pending", "dot"]);

      var itemClassName = (0, _classnames["default"])((_classNames = {}, (0, _defineProperty2["default"])(_classNames, "".concat(prefixCls, "-item"), true), (0, _defineProperty2["default"])(_classNames, "".concat(prefixCls, "-item-pending"), pending), _classNames), className);
      var dotClassName = (0, _classnames["default"])((_classNames2 = {}, (0, _defineProperty2["default"])(_classNames2, "".concat(prefixCls, "-item-head"), true), (0, _defineProperty2["default"])(_classNames2, "".concat(prefixCls, "-item-head-custom"), dot), (0, _defineProperty2["default"])(_classNames2, "".concat(prefixCls, "-item-head-").concat(color), true), _classNames2));
      return React.createElement("li", (0, _extends2["default"])({}, restProps, {
        className: itemClassName
      }), React.createElement("div", {
        className: "".concat(prefixCls, "-item-tail")
      }), React.createElement("div", {
        className: dotClassName,
        style: {
          borderColor: /blue|red|green/.test(color) ? undefined : color
        }
      }, dot), React.createElement("div", {
        className: "".concat(prefixCls, "-item-content")
      }, children));
    }
  }]);
  return TimelineItem;
}(React.Component);

exports["default"] = TimelineItem;
TimelineItem.defaultProps = {
  prefixCls: 'ant-timeline',
  color: 'blue',
  pending: false
};
