// antd/lib/layout/layout.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var React = _interopRequireWildcard(require("react"));

var PropTypes = _interopRequireWildcard(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var __rest = void 0 && (void 0).__rest || function (s, e) {
  var t = {};

  for (var p in s) {
    if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0) t[p] = s[p];
  }

  if (s != null && typeof Object.getOwnPropertySymbols === "function") for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
    if (e.indexOf(p[i]) < 0) t[p[i]] = s[p[i]];
  }
  return t;
};

function generator(props) {
  return function (BasicComponent) {
    return (
      /*#__PURE__*/
      function (_React$Component) {
        (0, _inherits2["default"])(Adapter, _React$Component);

        function Adapter() {
          (0, _classCallCheck2["default"])(this, Adapter);
          return (0, _possibleConstructorReturn2["default"])(this, (0, _getPrototypeOf2["default"])(Adapter).apply(this, arguments));
        }

        (0, _createClass2["default"])(Adapter, [{
          key: "render",
          value: function render() {
            var prefixCls = props.prefixCls;
            return React.createElement(BasicComponent, (0, _extends2["default"])({
              prefixCls: prefixCls
            }, this.props));
          }
        }]);
        return Adapter;
      }(React.Component)
    );
  };
}

var Basic =
/*#__PURE__*/
function (_React$Component2) {
  (0, _inherits2["default"])(Basic, _React$Component2);

  function Basic() {
    (0, _classCallCheck2["default"])(this, Basic);
    return (0, _possibleConstructorReturn2["default"])(this, (0, _getPrototypeOf2["default"])(Basic).apply(this, arguments));
  }

  (0, _createClass2["default"])(Basic, [{
    key: "render",
    value: function render() {
      var _a = this.props,
          prefixCls = _a.prefixCls,
          className = _a.className,
          children = _a.children,
          others = __rest(_a, ["prefixCls", "className", "children"]);

      var divCls = (0, _classnames["default"])(className, prefixCls);
      return React.createElement("div", (0, _extends2["default"])({
        className: divCls
      }, others), children);
    }
  }]);
  return Basic;
}(React.Component);

var BasicLayout =
/*#__PURE__*/
function (_React$Component3) {
  (0, _inherits2["default"])(BasicLayout, _React$Component3);

  function BasicLayout() {
    var _this;

    (0, _classCallCheck2["default"])(this, BasicLayout);
    _this = (0, _possibleConstructorReturn2["default"])(this, (0, _getPrototypeOf2["default"])(BasicLayout).apply(this, arguments));
    _this.state = {
      siders: []
    };
    return _this;
  }

  (0, _createClass2["default"])(BasicLayout, [{
    key: "getChildContext",
    value: function getChildContext() {
      var _this2 = this;

      return {
        siderHook: {
          addSider: function addSider(id) {
            _this2.setState(function (state) {
              return {
                siders: [].concat((0, _toConsumableArray2["default"])(state.siders), [id])
              };
            });
          },
          removeSider: function removeSider(id) {
            _this2.setState(function (state) {
              return {
                siders: state.siders.filter(function (currentId) {
                  return currentId !== id;
                })
              };
            });
          }
        }
      };
    }
  }, {
    key: "render",
    value: function render() {
      var _a = this.props,
          prefixCls = _a.prefixCls,
          className = _a.className,
          children = _a.children,
          hasSider = _a.hasSider,
          others = __rest(_a, ["prefixCls", "className", "children", "hasSider"]);

      var divCls = (0, _classnames["default"])(className, prefixCls, (0, _defineProperty2["default"])({}, "".concat(prefixCls, "-has-sider"), hasSider || this.state.siders.length > 0));
      return React.createElement("div", (0, _extends2["default"])({
        className: divCls
      }, others), children);
    }
  }]);
  return BasicLayout;
}(React.Component);

BasicLayout.childContextTypes = {
  siderHook: PropTypes.object
};
var Layout = generator({
  prefixCls: 'ant-layout'
})(BasicLayout);
var Header = generator({
  prefixCls: 'ant-layout-header'
})(Basic);
var Footer = generator({
  prefixCls: 'ant-layout-footer'
})(Basic);
var Content = generator({
  prefixCls: 'ant-layout-content'
})(Basic);
Layout.Header = Header;
Layout.Footer = Footer;
Layout.Content = Content;
var _default = Layout;
exports["default"] = _default;
